### ISSUE TITLE
Libraries linked more than once?

### ISSUE THREAD

**Issue by swingbit** [MonetDB/MonetDB#7367]
**Describe the bug**
Not a bug, possibly an improvement in the build scripts.

I am not a `cmake` expert, but I have the impression some of the `CMakeLists.txt`s could be improved.

All the `target_link_libraries` sections use the `PRIVATE` scope keyword for all items. I think this may in some cases result in the same libraries linked more than once.

For example:
```
target_link_libraries(sql
  PRIVATE
  monetdb_config_header
  sqlstorage
  sqlcommon
  sqlserver
  sqlinclude
  mcrypt
  batstore
  copybinary
  monetdb5
  bat)
```
could very well be:
```
target_link_libraries(sql
  PRIVATE
  monetdb_config_header
  sqlstorage
  sqlcommon
  sqlserver
  sqlinclude
  mcrypt
  batstore
  copybinary
  INTERFACE
  monetdb5
  bat)
```

How I noticed this: in my own module `spinque`, I used to have
```
target_link_libraries(spinque 
  PRIVATE
  monetdb_config_header
  mutils
  bat
  atoms
  monetdb5)
```

This would link a second copy of `atoms` to `spinque`, resulting in `atoms` being loaded twice, which triggers `STRprelude()` twice - which fails the second time, because it tries to create a second time the bats needed for case conversion. I have actually no idea why this failure didn't happen with Jan2022.

Changing the above to 
```
target_link_libraries(spinque 
  PRIVATE
  monetdb_config_header
  INTERFACE
  mutils
  bat
  atoms
  monetdb5)
```
fixed the issue, and the `spinque` library went from 2.5MB to 0.5MB (while staying fully functional).

**Software versions**
 - MonetDB version number 11.47.0
 - OS and version: Fedora 37
 - self-installed and compiled

**Comment by arjenderijke** [comment]
It is certainly possible that the CMake configuration could be improved.But there are several issues. First of all the version of cmake itself. Newer versions have improved the way dependencies are handled. So certain improvements might require a higher minimal cmake version. But that one might not be available in certain distributions.

The way details of how the cmake private/interface/public keywords work are not always obvious. To get everything working, you need a very strict definition of the library interfaces and dependencies. Given the fact that the MonetDB codebase is over 30 years old, it is not surprising that there were a lot of problems when i converted the Autotools configuration to CMake.

The biggest issue comes with transitive dependencies. That is where the difference between PRIVATE and INTERFACE really becomes important. If your code strictly defines a public interface for a library with a single header file that does not include other header files, then it is relatively easy. But in the MonetDB codebase there are a lot of places where for example header files include other header files from different libraries, leading to all kinds of unexpected dependencies. This makes it really difficult to clean up the cmake configuration without (major) changes to the code.

And it becomes more complicated when you consider optional features. When you compile MonetDB, you disable certain features, i assume. But certain issues might only arise when you enable a certain feature, let say the R integration. The definition you propose for the sql module might work in your case, but not when you enable the r-integration feature. And there are more than a dozen of these features.

And lets not forget the differences between different linux distributions and the compilation on Windows and MacOS. They all might lead to problems when defining a cmake configuration. And then there is the codebase of MonetDB itself. Changes to the code lead to different dependencies. That is likely the explanation of why this failure did not happen with the Jan2022 release.

To sum up, Cmake is a big improvement over Autotools. But it is not magic. And C compilation is just difficult :-) If you really want to improve the CMake configuration, you also need to improve the code itself.

**Comment by swingbit** [comment]
Hi @arjenderijke , yes I understand.

As I said I wasn't sure, but I suspected there could have been a somehow forgotten low-hanging fruit, so I brought this up.

I see now that if such fruit exists it's hanging much higher than I thought.

Cheers

**Comment by njnes** [comment]
some more (static) libs have been cleaned, ie less duplication.
